### DISCUSSION TITLE
URL LOGIN AND PWD

### DISCUSSION THREAD

**Discussion by gdiana** [grafana/grafana-infinity-datasource#312] [General]
Hi,

I have a case where I am accessing data for different clients, from the same datasource, but who use different LOGIN's and PWD's.

It appears currently the infinity data source URL setting allows only a single LOGIN and PWD to be used regardless of the client.

So I need to be able to somehow allow several LOGIN and PWD to be defined and to be able to parse a URL parameter that will insert the correct LOGIN and PWD for that client

ie    @client = 1 parse LOGIN=BOB&PWD= bob
ie    @client = 2 parse LOGIN=MARY&PWD=mary .... etc 

Is this currently possible and if not can it be done?

Otherwise, I have to expose each client's USERNAME and PASSWORD

Many thanks

Greg Diana

![image](https://user-images.githubusercontent.com/13445152/168323490-09077d2f-d3f4-4835-b407-0a919b2b4335.png)

**Comment by yesoreyeram** [comment]
You can't use that way. There are couple of other options. 

1. Create different datasource for different clients. And then make datasource as a variable.
2. Create username and password as dashboard variables. Like textbox and so you user can enter username and password in the variable textbox
3. Create a lookup table for username and password combination
But in my opinion none of the above are safe ways. So ideally you have to use one auth credentials for one datasource

**Reply by gdiana** [reply]
Thanks.

Hey by the way I have been playing with jsonata as it is very powerful.
I have resolved the tree issue by using the $. which allows one to retrieve stuff from higher up, and it works great.
I could not replicate your example but using jsonata $. I could.

The problems are, that I have found

1. all jsonata has to be on a single line
2. One cannot continue using 2 lines as jsonata bombs, SEE BELOW
3. 
![image](https://user-images.githubusercontent.com/13445152/168385460-814f85d0-7d5b-4f82-9946-874d41bc0604.png)

![image](https://user-images.githubusercontent.com/13445152/168385562-4846cb8f-857a-4e4d-923d-271db27bbc62.png)

I know you are probably aware, but the edit area is too small and needs may a pop out or drag enlarge, which is much easier to say than do.

But these are snippets as I go along.

1. Another thing I have gleened is that your jsonata is not a full implementation, but UQL with FULL jsonata, would be very powerful indeed.

Anyway, great work as I think you have opened up many opportunities and it is just people need to find and use them.

Do you intend expanding jsonata in terms of it many available functions. Below is what I have been playing with as an example.

Can it handle any function yet?

![image](https://user-images.githubusercontent.com/13445152/168386732-069a381b-020f-46c7-acf2-7cf3f02a525b.png)


Cheers

Greg Diana

**Reply by yesoreyeram** [reply]
Yea. Glad you solved it yourselves. ( I am not sure you shared the same xml i use or different xml .. previously you shared inline XML and then edited and pointed to some url .. so not sure if i using the correct one)

BTW, i am using the jsonata lib from  https://www.npmjs.com/package/jsonata so that must be full support i guess.. but not sure what is missing is library though .. 

I agree the editing experience can be improved.. will do few more iteratations.. mostly nowadays i am trying to fix critical bugs and get a stable release.  But any kind of feeback is really appreciated

**Reply by yesoreyeram** [reply]
Multi line string support is not included yet.. haven't found an easy way to fit into existing parser. Hopefully that will get improved over the time.

**Reply by gdiana** [reply]
The same package is used in NODE-RED but in NR they say you must 

![image](https://user-images.githubusercontent.com/13445152/168471748-d9eca814-d994-4e85-9088-5c538caa4b2c.png)

In NR for xml2json one has to do the following in the local js.settings files

![image](https://user-images.githubusercontent.com/13445152/168471879-2a10c75a-8435-4dde-ab8c-f276899cd239.png)

I am not sure how Grafana works but I presume it must be much the same.

**Reply by yesoreyeram** [reply]
It seems infinity supports functions as seen [here](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fjsonplaceholder.typicode.com%2Fusers%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22data%22:%22%5B%5D%22,%22uql%22:%22jsonata%20%5C%22(%20$pi%20:%3D%203.1415926535897932384626;%20$plot%20:%3D%20function($x)%20%7B(%20$floor%20:%3D%20$string%20~%3E%20$substringBefore(%3F,%20%27.%27)%20~%3E%20$number;%20$index%20:%3D%20$floor(($x%20%2B%201)%20*%2020%20%2B%200.5);$join(%5B0..$index%5D.(%27.%27))%20%26%20%27O%27%20%26%20$join(%5B$index..40%5D.(%27.%27)))%7D;%20$product%20:%3D%20function($a,%20$b)%20%7B%20$a%20*%20$b%20%7D;%20$factorial%20:%3D%20function($n)%20%7B%20$n%20%3D%200%20%3F%201%20:%20$reduce(%5B1..$n%5D,%20$product)%20%7D;%20$sin%20:%3D%20function($x)%7B%20$cos($x%20-%20$pi%2F2)%20%7D;%20$cos%20:%3D%20function($x)%7B%20$x%20%3E%20$pi%20%3F%20$cos($x%20-%202%20*%20$pi)%20:%20$x%20%3C%20-$pi%20%3F%20$cos($x%20%2B%202%20*%20$pi)%20:%20$sum(%5B0..12%5D.($power(-1,%20$)%20*%20$power($x,%202*$)%20%2F%20$factorial(2*$)))%7D;%20%5B0..24%5D.$sin($*$pi%2F12).$plot($))%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D). Not sure what else is missing.

**Reply by gdiana** [reply]
Ha, ha I suppose that resolves that issue.😂 🤣 😂 👍

**Comment by gdiana** [comment]
Ok I resolved the issue but maybe did not articulate matters correctly.

I simply created a second Infinity Datasource within Grafana with different logon details. Works perfectly.

What I meant was that the Clients use the same remote application Database from which I was reading data and not the Infinity Datasource.

So one may just use   Infinity Client1, Infinity Client  2......  and each may have different LOGON credentials
😀

**Reply by yesoreyeram** [reply]
If i understand correctly, that's how you are supposed to use it. ( Create different datasource instances and then create datasource variable type on your dashboard )
